src/validation.cpp

```diff
@@ -3741,6 +3741,102 @@ bool CVerifyDB::VerifyDB(CCoinsView* coinsview, int nCheckLevel, int nCheckDepth
     return true;
 }
 
+bool RewindBlockIndexToLastCheckpoint(const CChainParams& chainparams, bool& clearWitnessCaches, bool fJustCheck)
+{
+    LOCK(cs_main);
+
+    // Get current Height and current Checkpoints
+    int nHeight = chainActive.Height();
+    const CBlockIndex* prevCheckPoint = GetLastCheckpoint();
+    const int checkPointHeight = prevCheckPoint ? prevCheckPoint->nHeight : 0;
+
+    // First load options
+    if(!fJustCheck)
+        return true;
```

Comment:
Could we automatically detect if a rewind is needed?
If there are checkpoints between the tip and the last (known) checkpoint, we are not on the right chain.
Maybe something like this:

```
// If there are checkpoints between prevCheckPoint and chainActive.Tip
// This is not the right chain and a rewind to prevCheckPoint is needed
bool bNeedRewind = fJustCheck;
if (Checkpoints::fEnabled){
    const MapCheckpoints& checkpoints = *Params().Checkpoints().mapCheckpoints;

    for (const auto& i : reverse_iterate(checkpoints)) {
        if (i.first > checkPointHeight && i.first <= nHeight) {
            bNeedRewind = true;
            break;
        }
    }
}

if(!bNeedRewind)
    return true;
```

Reply:
If there are checkpoints between the tip and the last (known) checkpoint, we are not on the right chain. and this is right, 
but if there are not checkpoints between the tip and the last (known) checkpoint it does not necessarily mean we are on the right chain

Reply:
I do not see a possibility in automatically detecting a rewind.

In such we have two cases:
1.
There is a checkpoint known and in the future, but we are behind said checkpoint, we would then simply sync to it.
2.
There is not a checkpoint and between the tip and the last known, to which we should be fine.

IE: Checkpoint block is 2150000,  we are at block 2150010, if there was any in-between, that user would already have synced those
Checkpoint is block 2150005 and 2150000, while we are at block 2150010, there's no way we would be at 2150010 with a prevCheckpoint of 2150000 not 2150005 and on the same wallet version.